### PR TITLE
Add functionality to store and make available decay values of an Inventory

### PR DESCRIPTION
**Pull request by php1ic**
#### What does this PR implement/fix?
Start the work towards making the decayed data of an Inventory available. Fairly small commit to start in case the project is not interested in exploring this further.

#### Fixes Issue
N/A

#### Other comments?
Related to discussion #103

### CODE REVIEW

**Comment by alexmalins** [comment]
Hi Ian 👋 Thanks for contributing!

What is your vision for this & where do you plan on taking things?

The output of this new `calculate_decay_data()` method looks like something people could find useful in itself. But in itself it doesn't achieve what's discussed in #104. So I was wondering what your next mods will be?

Were you e.g. thinking of refactoring the `plot()` [method](https://github.com/radioactivedecay/radioactivedecay/blob/main/radioactivedecay/inventory.py#L939) to call `calculate_decay_data()` and reduce duplication of the similar code that now exists in both functions? I.e. to not use a `ydata` [numpy matrix](https://github.com/radioactivedecay/radioactivedecay/blob/main/radioactivedecay/inventory.py#L939) to transfer the y-data to the [function](https://github.com/radioactivedecay/radioactivedecay/blob/main/radioactivedecay/plots.py#L202) that actually performs the plotting, but instead pass this dict of nuclide strings and lists of floats?

I am happy in principle to incorporate this PR, just wondering on the direction you are thinking and where the stopping nuclides appearing in the legend if their line isn't visible on the graph will fit in.

**Comment by php1ic** [comment]
Sorry, should have laid out my plan for this first, the new function does seem a bit random and unrelated.

Definitely want to refactor the creation of the data out of the `plot()` function to then have it passed as a parameter after the user has updated or removed those nuclides they are not interested in. Might also be worth either adding the time data to the returned dictionary, or returning it separately as part of a tuple so the user doesn't have to recalculate when they want to plot?

Otherwise, how do you feel about starting to use [pandas](https://pandas.pydata.org/docs/index.html)? If the data is used to create a dataframe, it is simple to remove those nuclides that are never more active than a certain threshold.
```
>>> import radioactivedecay as rd
>>> import pandas as pd
>>> inv = rd.Inventory({"Rn-222": 10.0})
>>> dd = inv.calculate_decay_data(10, npoints=5)
>>> df = pd.DataFrame.from_dict(dd)
>>> print(df)
     At-218        Bi-210        Bi-214        Hg-206  Pb-206        Pb-210    Pb-214        Po-210        Po-214    Po-218        Rn-218     Rn-222        Tl-206        Tl-210
0  0.000000  0.000000e+00  2.399776e-15  0.000000e+00     0.0  0.000000e+00  0.000000  1.198314e-19 -2.078967e-15  0.000000  0.000000e+00  10.000000 -2.647370e-24 -2.802661e-19
1  0.000008 -3.307757e-18  2.818645e-08  0.000000e+00     0.0  2.408215e-17  0.000050  5.991569e-20  3.547652e-08  0.092732  7.305102e-09   9.999948 -2.647370e-24  3.298984e-14
2  0.000023 -3.307757e-18  2.184415e-07  1.065851e-26    -0.0  3.151620e-16  0.000199  5.991569e-20  2.406037e-07  0.184604  2.224014e-08   9.999895  2.647370e-24  5.111740e-13
3  0.000040  0.000000e+00  7.201571e-07  6.395105e-26     0.0  1.473199e-15  0.000447 -8.987354e-19  7.594014e-07  0.275623  3.946403e-08   9.999843  2.647370e-24  2.521978e-12
4  0.000058  0.000000e+00  1.676536e-06  2.451457e-25    -0.0  4.448915e-15  0.000792  1.198314e-19  1.733358e-06  0.365798  5.729271e-08   9.999790  2.647370e-24  7.801935e-12
>>> df_filtered = df.loc[:, df.max() > 1.0e-18]
>>> print(df_filtered)
     At-218        Bi-214        Pb-210    Pb-214        Po-214    Po-218        Rn-218     Rn-222        Tl-210
0  0.000000  2.399776e-15  0.000000e+00  0.000000 -2.078967e-15  0.000000  0.000000e+00  10.000000 -2.802661e-19
1  0.000008  2.818645e-08  2.408215e-17  0.000050  3.547652e-08  0.092732  7.305102e-09   9.999948  3.298984e-14
2  0.000023  2.184415e-07  3.151620e-16  0.000199  2.406037e-07  0.184604  2.224014e-08   9.999895  5.111740e-13
3  0.000040  7.201571e-07  1.473199e-15  0.000447  7.594014e-07  0.275623  3.946403e-08   9.999843  2.521978e-12
4  0.000058  1.676536e-06  4.448915e-15  0.000792  1.733358e-06  0.365798  5.729271e-08   9.999790  7.801935e-12
```

If we add the time data (as discussed above) as an entry to the dictionary/dataframe we could perhaps start using the built-in [plot function of dataframes](https://pandas.pydata.org/docs/getting_started/intro_tutorials/04_plotting.html)?

Apologies again, this post got a bit rambly once I started thinking about using pandas. However I do think a dataframe is the way to go. Once we have the calculated values in a dataframe, it can be filtered and plotted in almost any way the user requires.

What are your thoughts, are you happy for me to start storing the data in pandas dataframes and we can see how things unfold, or would you rather keep the data in dictionaries?

**Comment by alexmalins** [comment]
Hey Ian, I get your thinking and agree with using pandas. 

To fill some background, when I wrote this package originally was I tried to keep it lean and minimal on dependencies, as well as returning just basic python datatypes (dicts, strings, floats). This was to keep it very accessible to beginners, e.g. for use in education settings like schools.

With hindsight and with an eye on modernizing this package, it's going to be a better solution to be incorporating pandas moving forward. I had this in mind anyway for other future ideas, and it'll mean cleaner code under the hood.

My recs for this PR are:
- As you have let's offer users two ways to output the data using the two new functions - i.e. as a pandas dataframe and as a dict
- Under the hood it'll be cleaner to have the function that makes the data (i.e. calculates the decays) being the one that returns a df.
  - That'll be cleaner and more efficient code than the appending to the lists within the defaultdict method that `calculate_decay_data()` is currently using. 
  - You could turn the numpy matrix that `plot()` is using into the dataframe easily
  - Or you could initialize a df from the outset and write the data into it upon each `decay()` call
- The function returning a dict of lists could then be a simple call to the dataframe generating function, then returning `df.to_dict("list")`.
  - My preference for that one would be to return a tuple with the decay times list, and the dict with nuclides and lists, rather than having everything in the dict.


Either in this PR or a subsequent PR, `plot()` could then be refactored to plot from the dataframe, rather than the numpy matrix that is currently the case.

In order to get this PR ready to merge, please could you?
- Add pandas as a new dependency in requirements.txt and setup.cfg
- Document your new functions as appropriate. E.g. in the tutorial.rst and plotting.rst
- Add a changelog entry

Once those ducks are in line, I can review properly with inline comments 🚀

**Comment by php1ic** [comment]
It looks like creating the complete dictionary then using that to construct the dataframe is the simplest way to go. I also couldn't see a simple way to append to a dataframe in the way we require, and searching around, the consensus seemed to be that that was not the most efficient way to do it and using inbuilt containers with a single final conversion was best.

Pandas has been added as a dependency, and the documents updated.

I have purposely not yet altered the `plot()` function, but am happy to have a go and refactor the calculation of the decay out and replace with the new functionality.

File: docs/source/tutorial.rst [hunk 1]
```diff
@@ -126,6 +126,113 @@ picoseconds, nanoseconds, microseconds, milliseconds, seconds, minutes, hours,
...
+method runs ``decay()``, storing the data at each iteration and returning the
+complete data set as a pandas dataframe. The only required argument is the decay
+time period as a number, or as a numpy array with the individual decay times data
+is needed for. Optional arguments are the decay time units in the same format as for
+the ``decay()`` method, the decay units, whether the time scale should be linear or
+logarithmic and, if the decay time is given as a float, how many decay points should be
+calculated.
+
+Using ``calculate_decay_data()`` to interogate how the mass fraction of 14C decays over
+20,000 years with 14N taking it's place. The default value for the number of point to calculate
```
Comment (alexmalins) [inline docs/source/tutorial.rst:140]: You can use the following syntax if you'd prefer to have the 14's superscript:
```
 :sup:`14`\C 
 :sup:`14`\N
```
like [here](https://github.com/radioactivedecay/radioactivedecay/blob/main/docs/source/overview.rst?plain=1#L54)

File: docs/source/tutorial.rst [hunk 2]
```diff
@@ -126,6 +126,113 @@ picoseconds, nanoseconds, microseconds, milliseconds, seconds, minutes, hours,
...
+at the end of the time period, access to finer resolution decay data is required. The
+method runs ``decay()``, storing the data at each iteration and returning the
+complete data set as a pandas dataframe. The only required argument is the decay
+time period as a number, or as a numpy array with the individual decay times data
+is needed for. Optional arguments are the decay time units in the same format as for
+the ``decay()`` method, the decay units, whether the time scale should be linear or
+logarithmic and, if the decay time is given as a float, how many decay points should be
+calculated.
+
+Using ``calculate_decay_data()`` to interogate how the mass fraction of 14C decays over
```
Comment (alexmalins) [inline docs/source/tutorial.rst:139]: lil typo interogate -> interrogate

File: radioactivedecay/inventory.py [hunk 3]
```diff
@@ -828,6 +831,172 @@ def decay_modes(self) -> Dict[str, List[str]]:
...
+
+        Parameters
+        ----------
+        time_period : Union[float, np.ndarray]
+            Time to decay the chain for. If a float is given, <time_scale> and <npoints> is used to create an evenly spaced array of numbers.
+            If an numpy ndarray is provided, the values contained within are used and <npoints> is ignored.
+        time_units : str, optional
+            Units for time series. Options are 'ps', 'ns', 'μs', 'us', 'ms', 's', 'm', 'h', 'd', 'y',
+            'ky', 'My', 'By', 'Gy', 'Ty', 'Py', and common spelling variations. Default is 's', i.e.
+            seconds. Use 'readable' to get strings of the half-lives in human-readable units.
```
Comment (alexmalins) [inline radioactivedecay/inventory.py:854]: Delete `Use 'readable' to get strings of the half-lives in human-readable units.` as not used here 🙌

File: radioactivedecay/inventory.py [hunk 4]
```diff
@@ -828,6 +831,172 @@ def decay_modes(self) -> Dict[str, List[str]]:
...
+            Units to use for the decayed values e.g. 'Bq', 'kBq', 'Ci', 'g', 'mol', 'num',
+            'activity_frac', 'mass_frac', 'mol_frac'. Default is 'Bq'.
+        npoints : None or int, optional
+            Number of time points used to plot graph (default is 501 for normal precision decay
+            calculations, or 51 for high precision decay calculations).
+
+        Returns
+        -------
+        pandas.DataFrame
+            Pandas DataFrame with the data of the decayed Inventory. Each isotope is it's own column, with a row for
```
Comment (alexmalins) [inline radioactivedecay/inventory.py:867]: it's -> its I think, could be wrong 👍

File: radioactivedecay/inventory.py [hunk 5]
```diff
@@ -828,6 +831,172 @@ def decay_modes(self) -> Dict[str, List[str]]:
...
+
+        Parameters
+        ----------
+        time_period : Union[float, np.ndarray]
+            Time to decay the chain for. If a float is given, <time_scale> and <npoints> is used to create an evenly spaced array of numbers.
+            If an numpy ndarray is provided, the values contained within are used and <npoints> is ignored.
+        time_units : str, optional
+            Units for time series. Options are 'ps', 'ns', 'μs', 'us', 'ms', 's', 'm', 'h', 'd', 'y',
+            'ky', 'My', 'By', 'Gy', 'Ty', 'Py', and common spelling variations. Default is 's', i.e.
+            seconds. Use 'readable' to get strings of the half-lives in human-readable units.
```
Comment (alexmalins) [inline radioactivedecay/inventory.py:958]: ♻️

**Review by alexmalins — APPROVED** [review]
Super job on this 🙌 

I've only one suggestion which is to re-think about the naming of the new methods. `decayed_data()` and `calculate_decay_data()` do the same thing just different return types, and perhaps don't offer much distinction from the `decay()` method.

To be more descriptive would you consider naming them something like `decay_time_series()` and `decay_time_series_pandas()` instead?

What do you think? Open to suggestions as well.

> It looks like creating the complete dictionary then using that to construct the dataframe is the simplest way to go. I also couldn't see a simple way to append to a dataframe in the way we require, and searching around, the consensus seemed to be that that was not the most efficient way to do it and using inbuilt containers with a single final conversion was best.
>
> I have purposely not yet altered the plot() function, but am happy to have a go and refactor the calculation of the decay out and replace with the new functionality.

Understood. Now pandas is a dependency it gives us the option to offer proper scientific python return types (dfs or ndarrays) from the core Inventory methods in future. And at that point it may be possible to refactor the double for loops into shorter syntax by concatenating dfs or something.

Thanks again for this top notch contribution!

**Comment by php1ic** [comment]
Thanks, happy to help!

Strange how naming is often one of the most complicated parts of writing the code :laughing:. I like the descriptiveness and consistency of them all starting with `decay` so have gone with your suggestion.

The other comments have also been actioned.

**Comment by alexmalins** [comment]
🙌 thanks - please merge this then and I'll cut a new release

**Comment by php1ic** [comment]
Hi Alex,

I don't think I have permission to merge into main.

**Comment by alexmalins** [comment]
apologies - it'll be a repo setting somewhere but not sure which

a quick solution is for you to join the org if you're ok with that - just sent an invite

**Comment by alexmalins** [comment]
Hi Ian 👋 Sorry for the delay in cutting a new release for this - have been super busy with work and with a toddler cold moving family these last 10 days.
Aiming to get to it this weekend

**Comment by php1ic** [comment]
Hi Alex, No problem, whenever you have the time is fine :+1:

Hope the move has gone well and the toddler is feeling better.

**Pull request merged**